sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -695,22 +696,38 @@ ProgramManager::collectDependentDeviceImagesForVirtualFunctions(
       // virtual-functions-set properties, but their handling is the same: we
       // just grab all sets they reference and add them for consideration if
       // we haven't done so already.
+      bool isDummyImage = false;
       for (const sycl_device_binary_property &VFProp :
            BinImage->getVirtualFunctions()) {
-        std::string StrValue = DeviceBinaryProperty(VFProp).asCString();
+        if (VFProp->Name == std::string_view("dummy-image")) {
+          isDummyImage = true;
+          continue;
+        }
+        std::string_view StrValue = DeviceBinaryProperty(VFProp).asStringView();
         for (const auto &SetName : detail::split_string(StrValue, ',')) {
           if (HandledSets.insert(SetName).second)
             WorkList.push(SetName);
         }
       }
 
-      // TODO: Complete this part about handling of incompatible device images.
       // If device image uses the same virtual function set, then we only
       // link it if it is compatible.
       // However, if device image provides virtual function set and it is
       // incompatible, then we should link its "dummy" version to avoid link
       // errors about unresolved external symbols.
-      if (doesDevSupportDeviceRequirements(Dev, *BinImage))
+      // Note: we only link when exactly one of
+      // doesDevSupportDeviceRequirements(Dev, *BinImage) and
+      // isDummyImage is true. We don't want to link every dummy image,
+      // otherwise we could run into linking errors defining the same symbol
+      // multiple times. For every image providing virtual functions that has
+      // a dummy image, the dummy image will have the same device requirements
+      // as the original image. So when the dummy image does support the
+      // device requirements, we know that the corresponding image providing
+      // actual definitions will be linked and not the dummy. And vice versa:
+      // when the dummy image does not support the device requirements, we
+      // know the corresponding image providing virtual functions was not
+      // linked and we must link the dummy image.
+      if (doesDevSupportDeviceRequirements(Dev, *BinImage) + isDummyImage == 1)
```

Comment:
Since `isDummyImage` is a boolean and `doesDevSupportDeviceRequirements()` also returns a bool, I think it would be better to use XOR instead of add.
```suggestion
      if (doesDevSupportDeviceRequirements(Dev, *BinImage) ^ isDummyImage)
```